oss_src/unity/extensions/cumulative_aggregates.cpp

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (C) 2015 Dato, Inc.
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms
+ * of the BSD license. See the LICENSE file for details.
+ */
+#include <unity/lib/toolkit_function_macros.hpp>
+#include <unity/extensions/cumulative_aggregates.hpp>
+
+namespace graphlab {
+namespace cumulative_aggregates {
+
+template <typename AggregateFunctionType, 
+          typename AccumulatorType> 
+gl_sarray cumulative_aggregate(const gl_sarray& in,
+                     AggregateFunctionType aggregate_fn, 
+                     AccumulatorType init,
+                     flex_type_enum output_type) {
+  
+  // Empty case.  
+  size_t m_size = in.size();
+  if (m_size == 0) {
+    return gl_sarray({});
+  } 
+  
+  gl_sarray_writer writer(output_type, 1);
+  AccumulatorType y = init; 
+  for (const auto& v: in.range_iterator()) {
+    writer.write(aggregate_fn(v, y), 0); // z[t+1], y[t+1] = f(x[t], y[y]) 
```

Comment:
Is there any error handling that needs to be done for aggregate_fn? Must the aggregate_fn not throw?